tlair/src/main/scala/at/forsyte/apalache/tla/lir/transformations/LanguageWatchdog.scala

```diff
@@ -0,0 +1,33 @@
+package at.forsyte.apalache.tla.lir.transformations
+
+import at.forsyte.apalache.tla.lir.{TlaEx, TlaModule, UnexpectedLanguageError}
+
+/**
+  * Given a language predicate, the watchdog checks, whether an expression or a module satisfies the predicate.
+  * If not, the watchdog throws UnexpectedLanguageError.
+  *
+  * @param pred a language predicate
+  *
+  * @author Igor Konnov
+  */
+class LanguageWatchdog(pred: LanguagePred) {
+  private val predName: String = pred.getClass.getSimpleName
+
+  def check(ex: TlaEx): Unit = {
+    if (!pred.isExprOk(ex)) {
+      throw new UnexpectedLanguageError(s"Predicate $predName does not accept the expression: $ex")
+    }
+  }
+
+  def check(mod: TlaModule): Unit = {
+    if (!pred.isModuleOk(mod)) {
+      throw new UnexpectedLanguageError(s"Predicate $predName does not accept the module: ${mod.name}")
+    }
+  }
+}
+
+object LanguageWatchdog {
+  def apply(pred: LanguagePred): LanguageWatchdog = {
+    new LanguageWatchdog(pred)
+  }
+}
```

Comment:
case class + remove object?

Reply:
Case classes are used to defined abstract data types. Why making code more obfuscated instead of having 4 extra lines? Shorter code is not always better readable.

Reply:
It's not shorter if you now always define 
```
object X { 
  def apply(y) = new X(y) 
}
```

Reply:
If you define a case class, I would expect that you define another case class defined somewhere else. Using a sealed case class just for the purposes of defining a singleton is an abuse of the syntax.